tools/single_use/SaveUserIDToInstrumentData.php

```diff
@@ -0,0 +1,139 @@
+
+$confirm = false;
+if (isset($argv[1]) && $argv[1] === 'confirm') {
+    $confirm = true;
+}
+
+echo "\n\nQuerying data...\n\n";
+
+// Get history DB
+$DBInfo = $config->getSetting('database');
+$hDB = isset($DBInfo['historydb']) ? $DBInfo['historydb'] : $DBInfo['database'];
+
+$result_count = 1;
+
+// Query entries in the history table for the latest change
+// made to the flag table, for every CommentID.
+$history = $DB->pselect(
+    "SELECT f.Test_name, f.CommentID, h1.userID
+      FROM flag f
+      JOIN $hDB.history h1 on (f.CommentID=h1.primaryVals)
+      JOIN
+        (
+          SELECT primaryVals, MAX(changeDate) as changeDate
+            FROM $hDB.history
+            WHERE userID <> 'unknown'
+              AND type='U'
+            GROUP BY primaryVals
+        ) h2 USING (primaryVals, changeDate)
+            GROUP BY CommentID, h1.userID",
+    array()
+);
+// Loop and index results from history by testname
+$idxHist = [];
+foreach ($history as $entry) {
+    $idxHist[$entry['Test_name']][] = $entry;
+}
+
+foreach(\Utility::getAllInstruments() as $testname => $fullName) {
+    // Instantiate instrument object to get information
+    try {
+        $instrument = \NDB_BVL_Instrument::factory($testname, '', '');
+    } catch (Exception $e) {
+        echo "$testname does not seem to be a valid instrument.\n";
+        continue;
+    }
+
+    // Check if instrument saves data in JSON format i.e. no-SQL table
+    $JSONData = $instrument->usesJSONData();
+    $table = null;
+    if ($JSONData === false) {
+        $table = $instrument->table;
+        if (!$table) {
+            echo "\n\nERROR: The table name for instrument $testname cannot be found.\n";
+            continue;
+        } else if (!$DB->tableExists($table)) {
+            echo "Table $table for instrument $testname does not exist in the Database.\n";
+            continue;
+        }
+    }
+
+    if (empty($idxHist[$testname])) {
+        echo "\n\nThere is no data to import into $testname.\n";
+        continue;
+    } else {
+        echo "\n\nThe following data can be imported into $testname:\n\n";
+        foreach ($idxHist[$testname] as $row) {
+            $commentID = $row['CommentID'];
+            $userID = $row['userID'];
+            echo "\nResult $result_count: $commentID userID: $userID\n";
+            $result_count++;
+        }
+    }
+
+    // Update the instrument table
+    if ($confirm) {
+        echo "\nImporting data into $testname...\n\n";
+
+        // Instantiate instrument at the session level in order to
+        // use _saveValues() function
+        foreach ($idxHist[$testname] as $row) {
+            $commentID = $row['CommentID'];
+            $userID = $row['userID'];
+
+            try {
+                $sessionInst = \NDB_BVL_Instrument::factory(
+                    $testname,
+                    $commentID,
+                    ''
+                );
+            } catch (Exception $e) {
```

Comment:
is the only reason it would throw an exception that its not active ?

Reply:
i guess so? i took this bit of code from fix_candidate_age.php